arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/NonEmptySet.kt

```diff
@@ -54,14 +54,11 @@ public value class NonEmptySet<out A> private constructor(
 public fun <A> nonEmptySetOf(first: A, vararg rest: A): NonEmptySet<A> =
   NonEmptySet(first, rest.toSet())
 
-public fun <A> Iterable<A>.toNonEmptySetOrNull(): NonEmptySet<A>? =
-  firstOrNull()?.let { NonEmptySet(it, minus(it).toSet()) }
+public fun <A> Iterable<A>.toNonEmptySetOrNull(): NonEmptySet<A>? {
+  val iter = iterator()
+  if (!iter.hasNext()) return null
+  return NonEmptySet(iter.next(), Iterable { iter }.toSet())
+}
 
 public fun <A> Iterable<A>.toNonEmptySetOrNone(): Option<NonEmptySet<A>> =
   toNonEmptySetOrNull().toOption()
-
-public fun <A> Set<A>.toNonEmptySetOrNull(): NonEmptySet<A>? =
-  firstOrNull()?.let { NonEmptySet(it, minus(it)) }
-
-public fun <A> Set<A>.toNonEmptySetOrNone(): Option<NonEmptySet<A>> =
-  toNonEmptySetOrNull().toOption()
```

Comment:
Ah yes, I missed that this would cause a break.

Unfortunately this would not be faster - below is the implementation of `minus` for `Set` (I too miss Scala's structural sharing)

```kotlin
public operator fun <T> Set<T>.minus(element: T): Set<T> {
    val result = LinkedHashSet<T>(mapCapacity(size))
    var removed = false
    return this.filterTo(result) { if (!removed && it == element) { removed = true; false } else true }
}
```

however, given that `NonEmptySet` does not store its first element separately, unlike `NonEmptyList`, we could have a fast path for this by adding an `internal fun fromSetUnsafe` to the companion object and using that here in the case that the set is not empty?